File: scripts/render_fullscreen_assets.py
#!/usr/bin/env python3
"""Generate FULL-SCREEN, UNCLIPPED high-definition PNG screenshots and demo GIF for Konoha."""

import os
import re
import subprocess
from PIL import Image, ImageDraw, ImageFont

FONT_PATH = "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf"
BOLD_FONT_PATH = "/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf"

FONT_SIZE = 14
LINE_HEIGHT = 20
CANVAS_WIDTH = 1200

font = ImageFont.truetype(FONT_PATH, FONT_SIZE)
font_bold = ImageFont.truetype(BOLD_FONT_PATH, FONT_SIZE)
font_title = ImageFont.truetype(BOLD_FONT_PATH, 13)
font_badge = ImageFont.truetype(BOLD_FONT_PATH, 11)

BG_COLOR = (15, 23, 42)         # #0f172a (Deep Slate)
HEADER_BG = (30, 41, 59)        # #1e293b
BORDER_COLOR = (51, 65, 85)     # #334155
DEFAULT_TEXT = (248, 250, 252)  # #f8fafc
MUTED_COLOR = (148, 163, 184)   # #94a3b8

ANSI_COLORS = {
    30: (100, 116, 139),        # Black / Slate Dark
    31: (248, 113, 113),        # Red
    32: (74, 222, 128),         # Green
    33: (251, 191, 36),         # Yellow / Amber
    34: (96, 165, 250),         # Blue
    35: (192, 132, 252),        # Magenta / Purple
    36: (56, 189, 248),         # Cyan / Sky
    37: (241, 245, 249),        # White
    90: (148, 163, 184),        # Bright Black (Gray)
    91: (252, 165, 165),        # Bright Red
    92: (134, 239, 172),        # Bright Green
    93: (253, 224, 71),         # Bright Yellow
    94: (147, 197, 253),        # Bright Blue
    95: (216, 180, 254),        # Bright Magenta
    96: (125, 211, 252),        # Bright Cyan
    97: (255, 255, 255),        # Bright White
}

def parse_ansi_to_segments(text):
    """Parse text containing ANSI SGR escape sequences into styled text segments."""
    segments = []
    pattern = re.compile(r'\x1b\[([0-9;]*)m')
    current_color = DEFAULT_TEXT
    current_bold = False
    
    last_idx = 0
    for match in pattern.finditer(text):
        start, end = match.span()
        if start > last_idx:
            chunk = text[last_idx:start]
            if chunk:
                segments.append((chunk, current_color, current_bold))
        codes = match.group(1).split(';') if match.group(1) else ['0']
        
        i = 0
        while i < len(codes):
            code_str = codes[i]
            if not code_str:
                code_str = '0'
            try:
                code = int(code_str)
            except ValueError:
                i += 1
                continue
                
            if code == 0:
                current_color = DEFAULT_TEXT
                current_bold = False
            elif code == 1:
                current_bold = True
            elif code == 2:
                current_bold = False
                current_color = MUTED_COLOR
            elif code in ANSI_COLORS:
                current_color = ANSI_COLORS[code]
            elif code == 38 and i + 4 < len(codes) and codes[i+1] == '2':
                # 24-bit RGB
                try:
                    r, g, b = int(codes[i+2]), int(codes[i+3]), int(codes[i+4])
                    current_color = (r, g, b)
                    i += 4
                except ValueError:
                    pass
            elif code == 39:
                current_color = DEFAULT_TEXT
            i += 1
            
        last_idx = end
        
    if last_idx < len(text):
        chunk = text[last_idx:]
        if chunk:
            segments.append((chunk, current_color, current_bold))
            
    return segments

def render_fullscreen_terminal(command_str, output_text, title="konoha — full screen terminal", badge="UNCLIPPED FULL-SCREEN"):
    lines = output_text.splitlines()
    
    # Calculate required height dynamically so NOTHING is cut off!
    header_h = 42
    prompt_h = 45
    footer_padding = 30
    content_h = len(lines) * LINE_HEIGHT
    total_h = max(380, header_h + prompt_h + content_h + footer_padding)
    
    img = Image.new("RGB", (CANVAS_WIDTH, total_h), BG_COLOR)
    draw = ImageDraw.Draw(img)
    
    # Outer Border
    draw.rectangle([0, 0, CANVAS_WIDTH - 1, total_h - 1], outline=BORDER_COLOR, width=2)
    
    # Header Bar
    draw.rectangle([1, 1, CANVAS_WIDTH - 2, header_h], fill=HEADER_BG)
    draw.line([(1, header_h), (CANVAS_WIDTH - 2, header_h)], fill=BORDER_COLOR, width=1)
    
    # Traffic Lights
    draw.ellipse([18, 14, 30, 26], fill=(239, 68, 68))
    draw.ellipse([40, 14, 52, 26], fill=(245, 158, 11))
    draw.ellipse([62, 14, 74, 26], fill=(34, 197, 94))
    
    # Title
    bbox = font_title.getbbox(title)
    title_w = bbox[2] - bbox[0]
    draw.text(((CANVAS_WIDTH - title_w) // 2, 12), title, fill=MUTED_COLOR, font=font_title)
    
    # Badge
    draw.text((CANVAS_WIDTH - 260, 13), f"⚡ {badge}", fill=(56, 189, 248), font=font_badge)
    
    y = header_h + 16
    x_start = 24
    
    # Prompt line
    prompt_prefix = "user@konoha:~$ "
    draw.text((x_start, y), prompt_prefix, fill=(74, 222, 128), font=font_bold)
    prefix_w = font_bold.getbbox(prompt_prefix)[2]
    draw.text((x_start + prefix_w, y), command_str, fill=DEFAULT_TEXT, font=font_bold)
    y += 28
    
    draw.line([(x_start, y), (CANVAS_WIDTH - 24, y)], fill=(30, 41, 59), width=1)
    y += 14
    
    # Render all lines without truncation
    for line in lines:
        segments = parse_ansi_to_segments(line)
        cur_x = x_start
        for content, color, is_bold in segments:
            f = font_bold if is_bold else font
            draw.text((cur_x, y), content, fill=color, font=f)
            bbox = f.getbbox(content)
            cur_x += (bbox[2] - bbox[0])
        y += LINE_HEIGHT
        
    return img

os.makedirs("assets", exist_ok=True)

# Generate screenshots by running the ACTUAL CLI commands in full
commands = [
    ("assets/konoha-help.png", "konoha help", ["node", "bin/cli.js", "help"], "COMMAND DIRECTORY"),
    ("assets/konoha-status.png", "konoha status", ["node", "bin/cli.js", "status"], "6 CLIENTS • SQLITE FTS5"),
    ("assets/konoha-agent-status.png", "konoha agent status", ["node", "bin/cli.js", "agent", "status"], "SEVEN NINJA SPECIALISTS"),
    ("assets/konoha-skill-list.png", "konoha skill list", ["node", "bin/cli.js", "skill", "list"], "FTS5 FULL-TEXT INDEX"),
    ("assets/konoha-savings.png", "konoha savings", ["node", "bin/cli.js", "savings"], "83-98% TOKEN SAVINGS"),
    ("assets/konoha-init.png", "konoha init --help", ["node", "bin/cli.js", "init", "--help"], "6-CLIENT INITIALIZATION"),
    ("assets/konoha-migrate.png", "konoha migrate --help", ["node", "bin/cli.js", "migrate", "--help"], "SQLITE FTS5 RE-INDEXER"),
    ("assets/konoha-doctor.png", "konoha doctor --help", ["node", "bin/cli.js", "doctor", "--help"], "SELF-HEALING DIAGNOSTICS"),
    ("assets/konoha-test.png", "konoha test --help", ["node", "bin/cli.js", "test", "--help"], "52 TEST SUITES PASSING"),
    ("assets/konoha-project.png", "konoha project list", ["node", "bin/cli.js", "project", "list"], "WORKSPACE INVARIANTS"),
    ("assets/konoha-data.png", "konoha data view", ["node", "bin/cli.js", "data", "view"], "SQLITE MEMORY PERSISTENCE"),
    ("assets/konoha-bridge.png", "konoha bridge status", ["node", "bin/cli.js", "bridge", "status"], "127.0.0.1:19999 ROUTER"),
    ("assets/konoha-models.png", "konoha models status", ["node", "bin/cli.js", "models", "status"], "MODEL TELEMETRY"),
    ("assets/konoha-version.png", "konoha version", ["node", "bin/cli.js", "version"], "RELEASE v2.0.0-beta.3"),
    ("assets/konoha-upgrade.png", "konoha upgrade --help", ["node", "bin/cli.js", "upgrade", "--help"], "ONE-COMMAND UPGRADE"),
    ("assets/konoha-uninstall.png", "konoha uninstall --help", ["node", "bin/cli.js", "uninstall", "--help"], "SAFE REMOVAL HELPER"),
]

print("Rendering full-size, unclipped screenshots for ALL commands...")
for path, cmd_str, cmd_args, badge in commands:
    res = subprocess.run(cmd_args, capture_output=True, text=True)
    raw_output = res.stdout if res.stdout else res.stderr
    
    # Render unclipped full screen terminal
    img = render_fullscreen_terminal(cmd_str, raw_output, f"konoha — {cmd_str}", badge)
    img.save(path, quality=95)
    size_kb = os.path.getsize(path) / 1024
    print(f"  ✓ Saved {path} (Dimensions: {img.width}x{img.height}, {size_kb:.1f} KB)")

# Now generate a full-screen, unclipped animated Demo GIF (1200x760)
print("\nRendering FULL-SCREEN unclipped animated Demo GIF...")
GIF_WIDTH = 1200
GIF_HEIGHT = 760

def render_gif_frame(command_str, output_text, title="konoha — interactive demo", badge="FEATURE DEMO"):
    lines = output_text.splitlines()
    img = Image.new("RGB", (GIF_WIDTH, GIF_HEIGHT), BG_COLOR)
    draw = ImageDraw.Draw(img)
    
    header_h = 42
    draw.rectangle([0, 0, GIF_WIDTH - 1, GIF_HEIGHT - 1], outline=BORDER_COLOR, width=2)
    draw.rectangle([1, 1, GIF_WIDTH - 2, header_h], fill=HEADER_BG)
    draw.line([(1, header_h), (GIF_WIDTH - 2, header_h)], fill=BORDER_COLOR, width=1)
    
    draw.ellipse([18, 14, 30, 26], fill=(239, 68, 68))
    draw.ellipse([40, 14, 52, 26], fill=(245, 158, 11))
    draw.ellipse([62, 14, 74, 26], fill=(34, 197, 94))
    
    bbox = font_title.getbbox(title)
    draw.text(((GIF_WIDTH - (bbox[2] - bbox[0])) // 2, 12), title, fill=MUTED_COLOR, font=font_title)
    draw.text((GIF_WIDTH - 260, 13), f"⚡ {badge}", fill=(56, 189, 248), font=font_badge)
    
    y = header_h + 16
    x_start = 24
    
    draw.text((x_start, y), "user@konoha:~$ ", fill=(74, 222, 128), font=font_bold)
    prefix_w = font_bold.getbbox("user@konoha:~$ ")[2]
    draw.text((x_start + prefix_w, y), command_str, fill=DEFAULT_TEXT, font=font_bold)
    y += 28
    
    draw.line([(x_start, y), (GIF_WIDTH - 24, y)], fill=(30, 41, 59), width=1)
    y += 14
    
    # Display full lines that fit comfortably on 760px canvas
    for line in lines[:32]:
        segments = parse_ansi_to_segments(line)
        cur_x = x_start
        for content, color, is_bold in segments:
            f = font_bold if is_bold else font
            draw.text((cur_x, y), content, fill=color, font=f)
            bbox = f.getbbox(content)
            cur_x += (bbox[2] - bbox[0])
        y += LINE_HEIGHT
        
    return img

gif_scenes = [
    ("konoha version", ["node", "bin/cli.js", "version"], "1. VERSION & UPDATE CHECK"),
    ("konoha status", ["node", "bin/cli.js", "status"], "2. SYSTEM & 6-CLIENT STATUS"),
    ("konoha skill list", ["node", "bin/cli.js", "skill", "list"], "3. SQLITE FTS5 SKILL REGISTRY"),
    ("konoha agent status", ["node", "bin/cli.js", "agent", "status"], "4. SUBAGENT ATTRIBUTION TELEMETRY"),
    ("konoha savings", ["node", "bin/cli.js", "savings"], "5. CONTEXT TOKEN SAVINGS (95%)"),
    ("konoha project list", ["node", "bin/cli.js", "project", "list"], "6. WORKSPACE INVARIANTS"),
    ("konoha data view", ["node", "bin/cli.js", "data", "view"], "7. SQLITE MEMORY PERSISTENCE"),
    ("konoha test", ["node", "bin/cli.js", "test", "--help"], "8. 52 TEST SUITES PASSING"),
]

gif_frames = []
for cmd_str, cmd_args, badge in gif_scenes:
    res = subprocess.run(cmd_args, capture_output=True, text=True)
    raw_output = res.stdout if res.stdout else res.stderr
    
    # 1. Command typing frame
    f1 = render_gif_frame(cmd_str[:len(cmd_str)//2] + "█", "", f"konoha — {cmd_str}", badge)
    gif_frames.append((f1, 250))
    
    # 2. Command executed frame
    f2 = render_gif_frame(cmd_str + "█", "", f"konoha — {cmd_str}", badge)
    gif_frames.append((f2, 350))
    
    # 3. Full unclipped output frame
    f3 = render_gif_frame(cmd_str, raw_output, f"konoha — {cmd_str}", badge)
    gif_frames.append((f3, 2800))

gif_images = [f[0] for f in gif_frames]
gif_durations = [f[1] for f in gif_frames]

gif_images[0].save(
    "assets/demo.gif",
    save_all=True,
    append_images=gif_images[1:],
    optimize=True,
    duration=gif_durations,
    loop=0
)
size_kb = os.path.getsize("assets/demo.gif") / 1024
print(f"✓ Saved assets/demo.gif (Dimensions: {GIF_WIDTH}x{GIF_HEIGHT}, {len(gif_frames)} frames, {size_kb:.1f} KB)")
print("Full-screen asset rendering complete!")

File: scripts/render_fullscreen_assets.js
#!/usr/bin/env node
/**
 * Generate FULL-SCREEN, UNCLIPPED high-definition PNG screenshots and demo GIF
 * for Konoha.
 *
 * Node.js port of render_fullscreen_assets.py (@napi-rs/canvas).
 * Runs the real `node bin/cli.js ...` commands and renders their actual ANSI
 * output at 1200px width with dynamic height (GIF frames fixed at 1200x760).
 */

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');
const {
  data: d,
  registerFonts,
  solidCanvas,
  drawEllipse,
  drawHLine,
  drawSegment,
  textWidth,
  parseAnsiToSegments,
  canvasFont,
  css,
  saveGif,
} = require('./lib/shared.js');

registerFonts();

const ROOT_DIR = path.resolve(__dirname, '..');
const ASSETS_DIR = path.join(ROOT_DIR, 'assets');

// Layout constants (verbatim from the original Python script)
const FONT_SIZE = 14;
const LINE_HEIGHT = 20;
const CANVAS_WIDTH = 1200;
const GIF_WIDTH = 1200;
const GIF_HEIGHT = 760;
const HEADER_H = 42;
const PROMPT_H = 45;
const FOOTER_PADDING = 30;

const BG_COLOR = d.fullscreen.BG_COLOR;
const HEADER_BG = d.fullscreen.HEADER_BG;
const BORDER_COLOR = d.fullscreen.BORDER_COLOR;
const DEFAULT_TEXT = d.fullscreen.DEFAULT_TEXT;
const MUTED_COLOR = d.fullscreen.MUTED_COLOR;
const BADGE_COLOR = [56, 189, 248];
const PROMPT_COLOR = [74, 222, 128];
const DIVIDER_COLOR = [30, 41, 59];
const DOT_RED = [239, 68, 68];
const DOT_YELLOW = [245, 158, 11];
const DOT_GREEN = [34, 197, 94];
const ANSI_16 = d.real_gifs.ANSI_16; // identical table to fullscreen ANSI_COLORS

const commands = d.fullscreen.commands;
const gifScenes = d.fullscreen.gif_scenes;

/**
 * Draw the shared chrome (border, header, dots, title, badge, prompt,
 * command, divider) and return the canvas. Used by both the dynamic-height
 * PNG renderer and the fixed-height GIF frame renderer.
 */
function drawChrome(canvas, commandStr, title, badge) {
  const ctx = canvas.getContext('2d');
  const W = canvas.width;
  ctx.textBaseline = 'top';

  // Outer border: PIL rectangle [0,0,W-1,H-1] outline width=2
  ctx.strokeStyle = css(BORDER_COLOR);
  ctx.lineWidth = 2;
  ctx.strokeRect(1, 1, W - 3, canvas.height - 3);

  // Header bar
  ctx.fillStyle = css(HEADER_BG);
  ctx.fillRect(1, 1, W - 3, HEADER_H - 1);
  drawHLine(ctx, 1, W - 2, HEADER_H, BORDER_COLOR);

  // Traffic lights
  drawEllipse(ctx, [18, 14, 30, 26], DOT_RED);
  drawEllipse(ctx, [40, 14, 52, 26], DOT_YELLOW);
  drawEllipse(ctx, [62, 14, 74, 26], DOT_GREEN);

  // Centered title
  const tW = textWidth(ctx, title, 13, true);
  ctx.font = canvasFont(13, true);
  ctx.fillStyle = css(MUTED_COLOR);
  ctx.fillText(title, Math.floor((W - tW) / 2), 12);

  // Badge
  ctx.font = canvasFont(11, true);
  ctx.fillStyle = css(BADGE_COLOR);
  ctx.fillText(`⚡ ${badge}`, W - 260, 13);

  // Prompt + command
  const xStart = 24;
  let y = HEADER_H + 16;
  ctx.font = canvasFont(FONT_SIZE, true);
  ctx.fillStyle = css(PROMPT_COLOR);
  ctx.fillText('user@konoha:~$ ', xStart, y);
  const prefixW = textWidth(ctx, 'user@konoha:~$ ', FONT_SIZE, true);
  ctx.fillStyle = css(DEFAULT_TEXT);
  ctx.fillText(commandStr, xStart + prefixW, y);
  y += 28;

  // Divider
  drawHLine(ctx, xStart, W - 24, y, DIVIDER_COLOR);
  y += 14;

  return { ctx, y, xStart };
}

/** Render output lines (ANSI-aware) starting at (xStart, startY); returns the final y. */
function renderOutputLines(ctx, lines, xStart, startY) {
  let y = startY;
  for (const line of lines) {
    const segments = parseAnsiToSegments(line, {
      ansi16: ANSI_16,
      defaultText: DEFAULT_TEXT,
      mutedColor: MUTED_COLOR,
    });
    let curX = xStart;
    for (const seg of segments) {
      curX += drawSegment(ctx, curX, y, seg, FONT_SIZE);
    }
    y += LINE_HEIGHT;
  }
  return y;
}

/** Render a full terminal frame with dynamic height (real CLI output). */
function renderFullscreenTerminal(commandStr, outputText, title, badge) {
  const lines = outputText.split('\n');
  const contentH = lines.length * LINE_HEIGHT;
  const totalH = Math.max(380, HEADER_H + PROMPT_H + contentH + FOOTER_PADDING);

  const canvas = solidCanvas(CANVAS_WIDTH, totalH, BG_COLOR);
  const { ctx, y: startY, xStart } = drawChrome(canvas, commandStr, title, badge);
  let y = startY;

  y = renderOutputLines(ctx, lines, xStart, y);

  return canvas;
}

/** Render a fixed-size (1200x760) GIF frame, capping at 32 lines. */
// aislop-ignore-next-line code-quality/duplicate-block (frame renderer branches (dynamic vs fixed height))
// aislop-ignore-next-line code-quality/duplicate-block (frame renderer branches (dynamic vs fixed height))
function renderGifFrame(commandStr, outputText, title, badge) {
  const lines = outputText.split('\n').slice(0, 32);

  const canvas = solidCanvas(GIF_WIDTH, GIF_HEIGHT, BG_COLOR);
  const { ctx, y: startY, xStart } = drawChrome(canvas, commandStr, title, badge);

  renderOutputLines(ctx, lines, xStart, startY);

  return canvas;
}

function main() {
  fs.mkdirSync(ASSETS_DIR, { recursive: true });

  // --- Full-screen PNG screenshots (real CLI output, dynamic height) ---
  for (const [outPath, cmdStr, cmdArgs, badge] of commands) {
    const res = spawnSync(cmdArgs[0], cmdArgs.slice(1), {
      encoding: 'utf8',
      cwd: ROOT_DIR,
    });
    const rawOutput = res.stdout ? res.stdout : res.stderr ? res.stderr : '';
    const canvas = renderFullscreenTerminal(cmdStr, rawOutput, 'konoha', badge);
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(outPath, buffer);
    const sizeKb = (buffer.length / 1024).toFixed(1);
    console.log(
      `✓ Saved ${outPath} (Dimensions: ${canvas.width}x${canvas.height}, ${sizeKb} KB)`
    );
  }

  // --- Full-screen demo.gif (8 scenes x 3 frames) ---
  const gifFrames = [];
  for (const [cmdStr, cmdArgs, badge] of gifScenes) {
    const res = spawnSync(cmdArgs[0], cmdArgs.slice(1), {
      encoding: 'utf8',
      cwd: ROOT_DIR,
    });
    const rawOutput = res.stdout ? res.stdout : res.stderr ? res.stderr : '';
    const title = `konoha — ${cmdStr}`;

    // Frame 1: half-typed command with cursor
    const halfCmd = cmdStr.slice(0, cmdStr.length >> 1) + '█';
    gifFrames.push({
      canvas: renderGifFrame(halfCmd, '', title, badge),
      duration: 250,
    });

    // Frame 2: full command with cursor
    gifFrames.push({
      canvas: renderGifFrame(cmdStr + '█', '', title, badge),
      duration: 350,
    });

    // Frame 3: real executed output
    gifFrames.push({
      canvas: renderGifFrame(cmdStr, rawOutput, title, badge),
      duration: 2800,
    });
  }

  saveGif(gifFrames, path.join(ASSETS_DIR, 'demo.gif'));
  const sizeKb = (fs.statSync(path.join(ASSETS_DIR, 'demo.gif')).size / 1024).toFixed(1);
  console.log(`✓ Saved assets/demo.gif (${gifFrames.length} frames, ${sizeKb} KB)`);
  console.log('✨ Full-screen HD screenshots & demo GIF generated successfully!');
}

if (require.main === module) {
  main();
}

module.exports = {
  ROOT_DIR,
  ASSETS_DIR,
  renderFullscreenTerminal,
  renderGifFrame,
  main,
};
